nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -90,4 +132,141 @@ private static void invalidInstantiationError(
         errorBuilder.createErrorDescription(
             errorMessage, analysis.buildDescription(tree), state, null));
   }
+
+  private static void invalidAssignmentInstantiationError(
+      Tree tree,
+      String lhsAnnotation,
+      String rhsAnnotation,
+      VisitorState state,
+      NullAway analysis) {
+    ErrorBuilder errorBuilder = analysis.getErrorBuilder();
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.TYPE_PARAMETER_CANNOT_BE_NULLABLE,
+            String.format(
+                "Invalid Assignment, as assigning a %s value to a %s field",
+                rhsAnnotation, lhsAnnotation));
+    state.reportMatch(
+        errorBuilder.createErrorDescription(
+            errorMessage, analysis.buildDescription(tree), state, null));
+  }
+
+  public void checkInstantiationForAssignments(Tree tree) {
+    if (!config.isJSpecifyMode()) {
+      return;
+    }
+    Tree lhsTree;
+    Tree rhsTree;
+    if (tree instanceof VariableTree) {
+      VariableTree varTree = (VariableTree) tree;
+      lhsTree = varTree.getType();
+      rhsTree = varTree.getInitializer();
+    } else {
+      AssignmentTree assignmentTree = (AssignmentTree) tree;
+      lhsTree = assignmentTree.getVariable();
+      rhsTree = assignmentTree.getExpression();
+    }
+    // rhsTree can be null for a VariableTree.  Also, we don't need to do a check
+    // if rhsTree is the null literal
+    if (rhsTree == null || rhsTree.getKind().equals(Tree.Kind.NULL_LITERAL)) {
+      return;
+    }
+    Type lhsType = ASTHelpers.getType(lhsTree);
+    Type rhsType = ASTHelpers.getType(rhsTree);
+    if (rhsTree instanceof NewClassTree
+        && ((NewClassTree) rhsTree).getIdentifier() instanceof ParameterizedTypeTree) {
+      // for the parameterized typed tree ASTHelpers.getType() returns a type that does not have
+      // annotations preserved
+      rhsType =
+          typeWithPreservedAnnotations(
+              (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier());
+
+      ParameterizedTypeTree paramTypedTree =
+          (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier();
+      // not generic
+      if (paramTypedTree.getTypeArguments().size() <= 0) {
+        return;
+      }
+    }
+    if (lhsType != null && rhsType != null) {
+      compareAnnotations(lhsType, rhsType, tree);
+    }
+  }
+
+  private void compareAnnotations(Type lhsType, Type rhsType, Tree tree) {
+    if (lhsType instanceof Type.ClassType && rhsType instanceof Type.ClassType) {
+      rhsType = supertypeMatchingLHS((Type.ClassType) lhsType, (Type.ClassType) rhsType);
+    }
```

Comment:
Let's change the type of `lhsType` and `rhsType` parameters to be `ClassType` rather than doing this check; the method makes no sense if the types passed in are not `ClassType`s